pkg/services/oauthtoken/oauth_token.go

```diff
@@ -46,65 +59,130 @@ func (o *Service) GetCurrentOAuthToken(ctx context.Context, usr *user.SignedInUs
 	if err := o.AuthInfoService.GetAuthInfo(ctx, authInfoQuery); err != nil {
 		if errors.Is(err, user.ErrUserNotFound) {
 			// Not necessarily an error.  User may be logged in another way.
-			logger.Debug("no OAuth token for user found", "userId", usr.UserID, "username", usr.Login)
+			logger.Debug("no oauth token for user found", "userId", usr.UserID, "username", usr.Login)
 		} else {
-			logger.Error("failed to get OAuth token for user", "userId", usr.UserID, "username", usr.Login, "error", err)
+			logger.Error("failed to get oauth token for user", "userId", usr.UserID, "username", usr.Login, "error", err)
 		}
 		return nil
 	}
 
-	authProvider := authInfoQuery.Result.AuthModule
-	connect, err := o.SocialService.GetConnector(authProvider)
+	token, err := o.tryGetOrRefreshAccessToken(ctx, authInfoQuery.Result)
 	if err != nil {
-		logger.Error("failed to get OAuth connector", "provider", authProvider, "error", err)
 		return nil
 	}
+	return token
+}
+
+// IsOAuthPassThruEnabled returns true if Forward OAuth Identity (oauthPassThru) is enabled for the provided data source.
+func (o *Service) IsOAuthPassThruEnabled(ds *datasources.DataSource) bool {
+	return ds.JsonData != nil && ds.JsonData.Get("oauthPassThru").MustBool()
+}
+
+// HasOAuthEntry returns true and the UserAuth object when OAuth info exists for the specified User
+func (o *Service) HasOAuthEntry(ctx context.Context, usr *user.SignedInUser) (bool, *models.UserAuth) {
```

Comment:
This function should probably return (*models.UserAuth, error). We can then move the check of user auth existance further up and decide what to do in case of another error

Reply:
I'm not sure.`*models.UserAuth` can be nil (when there isn't any OAuth entry found for the user) and the caller would need multiple checks (one for the `err != nil` and another for `val != nil`) which I think is error-prone, because the second nil check can be easily forgotten. Besides, the method logs if there was an error fetching the data and the caller in the current context it cannot do anything with the error.

I would consider changing the result to `(bool, *models.UserAuth, error)` or rather to `(*models.UserAuth, bool, error)` to follow the 'comma ok' idiom, but it wouldn't add any value/improvement to the current implementation, because in case the GetAuthInfo call fails then the ExpiryDate check would be skipped anyway.

What do you think?

Reply:
Done